### PR TITLE
snapstate: move autorefresh code into autoRefresh helper

### PR DESCRIPTION
**Pull request by mvo5**
This cleans up the snapmgr.go file a bit and moves the
autorefresh logic into a single place/structure which
makes tests easier. Also adds some more tests around autoRefresh.

The first commit just moves things around and adds some tests,
the second commit adds using a new "RefreshScheduleManaged" helper.

Happy to split into two PRs if desired.

### CODE REVIEW

File: overlord/snapstate/snapmgr.go [hunk 1]
```diff
@@ -638,7 +469,7 @@ func (m *SnapManager) Ensure() error {
 		m.ensureForceDevmodeDropsDevmodeFromState(),
 		m.ensureUbuntuCoreTransition(),
 		m.refreshHints.Ensure(),
-		m.ensureRefreshes(),
```
Comment (zyga) [inline overlord/snapstate/snapmgr.go]: Nice pattern

File: overlord/snapstate/refreshhints_test.go [hunk 2]
```diff
@@ -62,7 +62,7 @@ func (s *refreshHintsTestSuite) SetUpTest(c *C) {
 	snapstate.CanAutoRefresh = func(*state.State) (bool, error) { return true, nil }
 }
 
-func (s *refreshHintsTestSuite) TearDownTests(c *C) {
```
Comment (zyga) [inline overlord/snapstate/refreshhints_test.go]: Oh, nice catch.

File: overlord/snapstate/autorefresh.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+// the default refresh pattern
+const defaultRefreshSchedule = "00:00-04:59/5:00-10:59/11:00-16:59/17:00-23:59"
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go]: we still need to decide what to do about this,  it's 5,6,6,7 hours but nor really a refactoring issue

File: overlord/snapstate/autorefresh.go [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+// hooks setup by devicestate
+var (
+	CanAutoRefresh         func(st *state.State) (bool, error)
+	RefreshScheduleManaged func(st *state.State) bool
+)
+
+// refreshRetryDelay specified the minimum time to retry failed refreshes
+var refreshRetryDelay = 10 * time.Minute
+
+func resetRefreshScheduleToDefault(st *state.State) (ts []*timeutil.Schedule, scheduleStr string) {
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go]: I would move this two helpers at the bottom of the file

File: overlord/snapstate/autorefresh.go (hunk 3 above)
Comment (mvo5) [reply overlord/snapstate/autorefresh.go]: I can fix it in this PR or in in a followup, whatever is easier.

File: overlord/snapstate/autorefresh.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	if err != nil {
+		return err
+	}
+
+	refreshSchedule, refreshScheduleStr, err := m.refreshScheduleWithDefaultsFallback()
+	if err != nil {
+		return err
+	}
+	if len(refreshSchedule) == 0 {
+		return nil
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go]: should we set m.nextRefresh = time.Time{}, there's no next refresh at this point

File: overlord/snapstate/autorefresh.go (hunk 3 above)
Comment (pedronis) [reply overlord/snapstate/autorefresh.go]: we could do it here, we need to think whether we want to do anything about old devices, but maybe for that it's best to wait the move from refresh.schedule to refresh.timer

Comment (mvo5) [reply overlord/snapstate/autorefresh.go]: Fixed

**Pull request merged**